src/services/completions.ts

```diff
@@ -3663,71 +3664,130 @@ function getCompletionData(
-            const firstAccessibleSymbolId = firstAccessibleSymbol && getSymbolId(firstAccessibleSymbol);
-            if (firstAccessibleSymbolId && addToSeen(seenPropertySymbols, firstAccessibleSymbolId)) {
-                const index = symbols.length;
-                symbols.push(firstAccessibleSymbol);
-                const moduleSymbol = firstAccessibleSymbol.parent;
-                if (
-                    !moduleSymbol ||
-                    !isExternalModuleSymbol(moduleSymbol) ||
-                    typeChecker.tryGetMemberInModuleExportsAndProperties(firstAccessibleSymbol.name, moduleSymbol) !== firstAccessibleSymbol
-                ) {
-                    symbolToOriginInfoMap[index] = { kind: getNullableSymbolOriginInfoKind(SymbolOriginInfoKind.SymbolMemberNoExport) };
+        if (!computedPropertyName) {
+            addLiteralSymbol();
+            return;
+        }
+
+        const computedPropertyNameExpression = computedPropertyName.expression;
+        const name = isEntityName(computedPropertyNameExpression) ? computedPropertyNameExpression :
+            isPropertyAccessExpression(computedPropertyNameExpression) ? computedPropertyNameExpression.name : undefined;
+        const nameSymbol = name && typeChecker.getSymbolAtLocation(name);
+        const nameSymbolId = nameSymbol && getSymbolId(nameSymbol);
+        if (!nameSymbolId) { // Not a property access or entity name
+            addLiteralSymbol();
+            return;
+        }
+
+        if (addToSeen(seenPropertySymbols, nameSymbolId)) {
+            const leftMostName = getLeftMostName(computedPropertyNameExpression);
+            const leftMostNameSymbol = leftMostName && typeChecker.getSymbolAtLocation(leftMostName);
+            const firstAccessibleSymbol = leftMostNameSymbol && getFirstSymbolInChain(leftMostNameSymbol, contextToken, typeChecker);
+            if (!firstAccessibleSymbol) { // Symbol is not accessible from completion location
+                addLiteralSymbol();
+                return;
+            }
+
+            const index = symbols.length;
+            symbols.push(nameSymbol);
+            const moduleSymbol = firstAccessibleSymbol.parent;
+            if (
+                !moduleSymbol ||
+                !isExternalModuleSymbol(moduleSymbol) ||
+                typeChecker.tryGetMemberInModuleExportsAndProperties(firstAccessibleSymbol.name, moduleSymbol) !== firstAccessibleSymbol
+            ) {
+                // If preferences allow insert text, add completion for [<QualifiedSymbolName>]
+                const node = preferences.includeCompletionsWithInsertText
+                    ? createComputedPropertyAccess(nameSymbol, leftMostNameSymbol, computedPropertyNameExpression)
+                    : undefined;
+
+                if (!node) {
+                    // Switch to literal symbol if user doesn't want insert text
+                    symbols[index] = symbol;

```

Comment:
Could we move `symbols.push(nameSymbol)` to the `else` below and call `addLiteralSymbol()` here instead? I think it would be clearer this way, and also make sure we're still setting the things `addLiteralSymbol` sets (i.e. the correct sort text).